[FLINK-3226] Implement a CodeGenerator for an efficient translation to DataSet programs

twalthr opened the pull request:
This PR implements a CodeGenerator for Table API on Calcite.

It includes the following:
- FilterITCase, SelectITCase and ExpressionITCase are mostly working again
- Null values are now fully supported and tested with new `TableProgramsTestBase`
- Arithmetic `+-*/ +X -X`, logical operators `== != AND, OR, NOT, IS NULL, NOT NULL` are supported.
- Logical operators implement 3-valued logic.
- A new configuration parameter `efficientTypeUsage` allows generated DataSet programs to be as efficient as ordinary DataSet programs by avoiding unnecessary type conversions and using most efficient type for every operator.

Limitations:
- String functions are not yet supported. They will be forwarded to Calcites runtime functions once I have implemented the necessary interfaces for that.
- Expression type casting is missing yet
- Date Literal is not supported yet

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/codegen/CodeGenerator.scala:
```diff
@@ -0,0 +1,661 @@
+          throw new CodeGenException("Decimal can not be converted to double.")
+        }
+      case VARCHAR | CHAR =>
+        generateNonNullLiteral(resultType, value.toString)
```
Do we need quotes around the string value?

twalthr replied:
Yes, this is a bug. Thanks.

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/codegen/CodeGenerator.scala:
```diff
@@ -0,0 +1,661 @@
+import org.apache.flink.api.scala.typeutils.CaseClassTypeInfo
+import org.apache.flink.api.table.TableConfig
+import org.apache.flink.api.table.codegen.CodeGenUtils._
+import org.apache.flink.api.table.codegen.Indenter._
```
Intellij tells me this import is unused

twalthr replied:
I had the same problem with IntelliJ. But if you remove the import, the `j"..."` will not work anymore. Actually I think we don't need the Indenter at all.

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/codegen/CodeGenUtils.scala:
```diff
@@ -0,0 +1,176 @@
+        ObjectFieldAccessor("f" + index)
+
+      case pj: PojoTypeInfo[_] =>
+        ObjectFieldAccessor(pj.getFieldNames()(index))
```
PojoFields may be private. 
You need to use reflection and previously make the field accessible.

twalthr replied:
I forgot to mention this in the limitations list. I will implement a case distinction and call the getters for private fields in my next commit.

fhueske replied:
The PojoSerializer and PojoComparator always use reflection to access the fields to avoid any side effects when calling the getters and setters.

fhueske commented:
Hi Timo, the PR looks really good :-)

I found the following issues / questions:
- Accessing of POJO fields might not work.
- Can you add method comments to the code generation methods in `CodeGenerator` and `CodeGenUtils`?
- Would it make sense to separate the function and expression code gen, i.e., split the `CodeGenerator` class?

twalthr commented:
Thanks for reviewing @fhueske.

I will comment my code more.

Regarding the splitting of the `CodeGenerator`, I think it is not that easy because functions and expressions both access reusable code parts. I will think about it again...

twalthr commented:
I have added comments and fixed the string bug. I will solve the other issues at the end of the week.
If you have no objections, you can merge the PR ;-)

vasia commented:
Thanks for the quick update @twalthr! Some tests are failing because the wrong type of exception is expected. I'll fix those and then merge this.

vasia commented:
Hey @twalthr,
I merged this yesterday but the PR didn't automatically close. Could you please close it manually? Thanks!
